solr/core/src/java/org/apache/solr/cli/StreamTool.java

```diff
@@ -0,0 +1,507 @@
+    String execution = cli.getOptionValue(EXECUTION_OPTION,"solr");
+    String arrayDelimiter = cli.getOptionValue(ARRAY_DELIMITER_OPTION,"|");
+    String delimiter = cli.getOptionValue(DELIMITER_OPTION, "   ");
+    boolean includeHeaders = cli.hasOption(HEADER_OPTION);
+    String[] outputHeaders = getOutputFields(cli);
+
+    LineNumberReader bufferedReader = null;
+    String expr;
+    try {
+      Reader inputStream =
+          expressionArgument.toLowerCase(Locale.ROOT).endsWith("expr")
+              ? new InputStreamReader(
+                  new FileInputStream(expressionArgument), Charset.defaultCharset())
+              : new StringReader(expressionArgument);
+
+      bufferedReader = new LineNumberReader(inputStream);
+      expr = StreamTool.readExpression(bufferedReader, cli.getArgs());
+      echoIfVerbose("Running Expression: " + expr);
+    } finally {
+      if (bufferedReader != null) {
+        bufferedReader.close();
+      }
+    }
+
+    PushBackStream pushBackStream;
+    if (execution.equalsIgnoreCase("local")) {
+      pushBackStream = doLocalMode(cli, expr);
+    } else {
+      pushBackStream = doRemoteMode(cli, expr);
+    }
+
+    try {
+      pushBackStream.open();
+
+      if (outputHeaders == null) {
+
+        Tuple tuple = pushBackStream.read();
+
+        if (!tuple.EOF) {
+          outputHeaders = getHeadersFromFirstTuple(tuple);
+        }
+
+        pushBackStream.pushBack(tuple);
+      }
+
+      if (includeHeaders) {
+        StringBuilder headersOut = new StringBuilder();
+        if (outputHeaders != null) {
+          for (int i = 0; i < outputHeaders.length; i++) {
+            if (i > 0) {
+              headersOut.append(delimiter);
+            }
+            headersOut.append(outputHeaders[i]);
+          }
+        }
+        CLIO.out(headersOut.toString());
+      }
+
+      while (true) {
+        Tuple tuple = pushBackStream.read();
+        if (tuple.EOF) {
+          break;
+        } else {
+          StringBuilder outLine = new StringBuilder();
+          if (outputHeaders != null) {
+            for (int i = 0; i < outputHeaders.length; i++) {
+              if (i > 0) {
+                outLine.append(delimiter);
+              }
+
+              Object o = tuple.get(outputHeaders[i]);
+              if (o != null) {
+                if (o instanceof List) {
+                  List outfields = (List) o;
+                  outLine.append(listToString(outfields, arrayDelimiter));
+                } else {
+                  outLine.append(o);
+                }
+              }
+            }
+          }
+          CLIO.out(outLine.toString());
+        }
+      }
+    } finally {
+
+      if (pushBackStream != null) {
+        pushBackStream.close();
+      }
+
+      solrClientCache.close();
+    }
+
+    echoIfVerbose("StreamTool -- Done.");
+  }
+
+  public PushBackStream doLocalMode(CommandLine cli, String expr) throws Exception {
```

Comment:
I believe these methods could all be private if the tests do not directly access them.